### PR TITLE
Sort by recently used

### PR DESCRIPTION
**Pull request by Dioxo**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## :scroll: Description

I added a new `PasswordSortOrder` for recently used passwords, saving the time when the password was used and comparing with others.

## :bulb: Motivation and Context

#535 

## :pencil: Checklist

- [x] I formatted the code with the IDE's reformat action (Ctrl + Shift + L/Cmd + Shift + L)
- [x] I reviewed submitted code
- [x] I added a [CHANGELOG](CHANGELOG.md) entry if applicable

### CODE REVIEW

File: app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt:251]: @msfjarvis Would it be better to always save the timestamp? I'm torn.

File: app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt]: Please use a separate shared preferences file here as well as the Kotlin `edit` extension.

File: CHANGELOG.md
Comment (fmeum) [inline CHANGELOG.md]: Let's keep the original format.

**Review by fmeum — COMMENTED** [review]
Thanks for the v2, this approach looks pretty good to me. Just a few comments.

File: CHANGELOG.md
Comment (msfjarvis) [inline CHANGELOG.md]: Create a "Added" headed under unreleased and add this there, reset all the other changes.

File: app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt
Comment (msfjarvis) [reply app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt:251]: Reading is quicker than writing so it's probably best to avoid a SharedPrefs edit when possible.

**Review by msfjarvis — COMMENTED** [review]
Seems good to me, address Fabian's review comments and I'll merge it after testing. Great job!

File: app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt
Comment (Dioxo) [reply app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt:251]: Well, if we always save the timestamp, when this option is selected then everything will be sorted (and so probably  passwords won't be were they used to be... and could be a little weird to find at first), so i thought to keep an Independent sort at first and then start keeping track of the recent passwords.

File: app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt
Comment (Dioxo) [reply app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt]: I'll do it

**Comment by Dioxo** [comment]
would it be better to save the file path or a hash?
and I also think that if the password/category is modified, then the previous name is saved and it is not modified. I'm going to fix that too.

File: app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt
Comment (fmeum) [reply app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt:251]: Agreed, let's keep this as is.

**Comment by Dioxo** [comment]
I think i included all your reviews and manage the case of renaming category/password
tell me what you think now :) @msfjarvis @FabianHenneke

File: CHANGELOG.md
Comment (fmeum) [inline CHANGELOG.md]:
```suggestion
```

File: app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt]:
```suggestion
                        val preferences = context.getSharedPreferences("recent_password_history", Context.MODE_PRIVATE)
```

File: app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/PasswordFragment.kt]: This will also need an import.
```suggestion
                        preferences.edit {
                            putString(item.file.absolutePath, System.currentTimeMillis().toString())
                        }
```

File: app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt]: See above, please use the Kotlin `edit` extension.

File: app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:761]: I don't know the exact charset supported by preference keys. @msfjarvis Do you know whether arbitrary paths could cause an issue here? Base64 or a hash might be safer.

File: app/src/main/java/com/zeapo/pwdstore/crypto/PasswordCreationActivity.kt
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/crypto/PasswordCreationActivity.kt]: See above

File: app/src/main/java/com/zeapo/pwdstore/utils/PasswordRepository.kt
Comment (fmeum) [inline app/src/main/java/com/zeapo/pwdstore/utils/PasswordRepository.kt:40]: This could be simplified using `compareBy` and `nullsLast`, but I could also add that later.

**Review by fmeum — COMMENTED** [review]
Some more comments, but I'm overall quite happy. Thanks for the contribution!

File: app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt
Comment (msfjarvis) [reply app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:761]: > I don't know the exact charset supported by preference keys. @msfjarvis Do you know whether arbitrary paths could cause an issue here? Base64 or a hash might be safer.

It's probably UTF_8 but a base64 hash is the safest bet.

Comment (Dioxo) [reply app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:761]: So... should I change the path to a hash...? @msfjarvis @FabianHenneke

File: app/src/main/java/com/zeapo/pwdstore/utils/PasswordRepository.kt
Comment (Dioxo) [reply app/src/main/java/com/zeapo/pwdstore/utils/PasswordRepository.kt:40]: I'll leave it to you, I have no idea how to use those...

File: app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt
Comment (msfjarvis) [reply app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:761]: Yes please

Comment (Dioxo) [reply app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:761]: Okay, i'll do it tomorrow :)

Comment (msfjarvis) [reply app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:761]: I had some time so I did it, hope you don't mind.

Comment (Dioxo) [reply app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:761]: Great :)

**Review by msfjarvis — APPROVED** [review]

File: app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt
Comment (Dioxo) [inline app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:761]: We should update this path to a base64 too. @msfjarvis

**Review by Dioxo — COMMENTED** [review]
Hey, @msfjarvis I think you missed one here too.

File: app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt
Comment (msfjarvis) [reply app/src/main/java/com/zeapo/pwdstore/PasswordStore.kt:761]: Oh weird, I definitely remember going back and fixing this one. Will push another PR in a bit.

**Pull request merged**
